examples/step-66/step-66.cc

```diff
@@ -0,0 +1,1120 @@
+        if (ERRf < TOLf || ERRx < TOLx)
+          {
+            solver_timer.stop();
+
+            pcout << "Convergence step " << newton_step << " value " << ERRf
+                  << " (used wall time: " << solver_timer.wall_time() << " s)"
+                  << std::endl;
+
+            break;
+          }
+        else if (newton_step == itmax)
+          {
+            solver_timer.stop();
+            pcout << "WARNING: No convergence of Newton's method after "
+                  << newton_step << " steps." << std::endl;
+
+            break;
+          }
+      }
+  }
+
+
+
+  // @sect4{GelfandProblem::compute_solution_norm}
+
+  // The computation of the H1-seminorm of the solution can be done in the
+  // same way as in step-59. We update the ghost values and use the function
+  // integrate_difference from the VectorTools namespace. In the end we gather
+  // all computations from all MPI ranks and return the norm.
+  template <int dim, int fe_degree>
+  double GelfandProblem<dim, fe_degree>::compute_solution_norm() const
+  {
+    solution.update_ghost_values();
+
+    Vector<float> norm_per_cell(triangulation.n_active_cells());
+
+    VectorTools::integrate_difference(mapping,
+                                      dof_handler,
+                                      solution,
+                                      Functions::ZeroFunction<dim>(),
+                                      norm_per_cell,
+                                      QGauss<dim>(fe.degree + 2),
+                                      VectorTools::H1_seminorm);
+
+    return VectorTools::compute_global_error(triangulation,
+                                             norm_per_cell,
+                                             VectorTools::H1_seminorm);
+  }
+
+
+
+  // @sect4{GelfandProblem::output_results}
+
+  // We generate the graphical output files in vtu format together with a pvtu
+  // master file in the same way as in step-37. Note that we also write the
+  // distribution of the triangulation into the output file as it was done in
+  // step-40. So no further comments are necessary.
+  template <int dim, int fe_degree>
+  void
+  GelfandProblem<dim, fe_degree>::output_results(const unsigned int cycle) const
+  {
+    if (triangulation.n_global_active_cells() > 1e6)
+      return;
+
+    solution.update_ghost_values();
+
+    DataOut<dim> data_out;
+    data_out.attach_dof_handler(dof_handler);
+    data_out.add_data_vector(solution, "solution");
+
+    Vector<float> subdomain(triangulation.n_active_cells());
+    for (unsigned int i = 0; i < subdomain.size(); ++i)
+      {
+        subdomain(i) = triangulation.locally_owned_subdomain();
+      }
+    data_out.add_data_vector(subdomain, "subdomain");
+
+    data_out.build_patches(mapping,
+                           fe.degree); // TODO coarse meshes look strange in
+                                       // paraview if we give the mapping object
```

Comment:
I would not be surprised that need to specify `DataOut::CurvedCellRegion::curved_inner_cells`.

Reply:
Yes, you are absolutely right, see the discussion in gfcas#4 with @marcfehling ;-)